### ISSUE TITLE
Rule#each. Undefined method `key?' for nil:NilClass

### ISSUE THREAD

**Issue by vatnajokull** [dry-rb/dry-validation#670]
**Describe the bug**

After calling contract with nested hash when the root key is nil and rule with `.each` applied the error occurs
```
NoMethodError:
   undefined method `key?' for nil:NilClass
```
**To Reproduce**
Giving a simple contract like this one:

```
class Index < Dry::Validation::Contract
  params do
    optional(:filter).hash do
      optional(:user_ids).array(:integer)
    end
   end

  rule(filter: :user_ids).each do
    key.failure('Not found') unless User.exists?(value)
  end
end

Index.new.call(filter: nil)
```
**Expected behavior**

**Expected behavior**
The contract does not raise any errors.

**Your environment**
- dry-validation 1.5.6
- Ruby 2.6.5

**Comment by solnic** [comment]
this is actually expected, since you defined `filter.user_ids[]` as an optional key, it may or may not exist, thus its value may not exist. You need to add an additional check to your rule, ie:

```ruby
class Index < Dry::Validation::Contract
  params do
    optional(:filter).hash do
      optional(:user_ids).array(:integer)
    end
   end

  rule(filter: :user_ids).each do
    if key?
      key.failure('Not found') unless User.exists?(value)
    end
  end
end

Index.new.call(filter: nil)
```

I wonder if this could be automated though... 🤔 After all, we can check that a given key is optional or that a value is defined as "maybe", then we could use that information to automatically add a rule guard (essentially that `key?` check).

I'll give this a shot and report back.
